File: NewsDaily/app/src/main/java/com/andruid/magic/newsdaily/pref/CountryPreferenceDialogFragment.java
package com.andruid.magic.newsdaily.pref;

import android.os.Bundle;
import android.view.View;

import androidx.preference.DialogPreference;
import androidx.preference.PreferenceDialogFragmentCompat;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.andruid.magic.newsdaily.adapter.CountryAdapter;
import com.andruid.magic.newsdaily.databinding.PrefDialogCountryBinding;
import com.andruid.magic.newsdaily.eventbus.CountryEvent;
import com.andruid.magic.newsdaily.util.AssetsUtil;
import com.blongho.country_data.Country;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;

import java.io.IOException;
import java.util.List;
import java.util.Objects;

import timber.log.Timber;

public class CountryPreferenceDialogFragment extends PreferenceDialogFragmentCompat {
    private CountryAdapter countryAdapter;
    private String country;

    public static CountryPreferenceDialogFragment newInstance(String key){
        CountryPreferenceDialogFragment fragment = new CountryPreferenceDialogFragment();
        final Bundle b = new Bundle(1);
        b.putString(ARG_KEY, key);
        fragment.setArguments(b);
        return fragment;
    }

    @Override
    public void onResume() {
        super.onResume();
        EventBus.getDefault().register(this);
    }

    @Override
    public void onPause() {
        super.onPause();
        EventBus.getDefault().unregister(this);
    }

    @Override
    public void onDialogClosed(boolean positiveResult) {
        if(positiveResult && country != null){
            DialogPreference preference = getPreference();
            if(preference instanceof CountryPreference){
                ((CountryPreference)preference).setCountry(country);
            }
        }
    }

    @Override
    protected void onBindDialogView(View view) {
        super.onBindDialogView(view);
        PrefDialogCountryBinding binding = PrefDialogCountryBinding.bind(view);
        if(binding == null)
            return;
        try {
            List<Country> countries = AssetsUtil.getCountries(Objects.requireNonNull(getContext())
                    .getAssets());
            countryAdapter = new CountryAdapter(countries);
            binding.recyclerView.setAdapter(countryAdapter);
            binding.recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
            binding.recyclerView.setItemAnimator(new DefaultItemAnimator());
            binding.recyclerView.addItemDecoration(new DividerItemDecoration(getContext(),
                    DividerItemDecoration.VERTICAL));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Subscribe
    public void onCountryEvent(CountryEvent countryEvent){
        country = countryEvent.getCountryCode();
        Timber.tag("preflog").d("country clicked: %s", country);
        Objects.requireNonNull(getDialog()).dismiss();
        onDialogClosed(true);
    }
}

File: NewsDaily/app/src/main/java/com/andruid/magic/newsdaily/pref/CountryPreferenceDialogFragment.kt
package com.andruid.magic.newsdaily.pref

import android.util.Log
import android.view.View
import androidx.core.os.bundleOf
import androidx.preference.PreferenceDialogFragmentCompat
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import com.andruid.magic.newsdaily.ui.adapter.CountryAdapter
import com.andruid.magic.newsdaily.databinding.PrefDialogCountryBinding
import com.andruid.magic.newsdaily.eventbus.CountryEvent
import com.andruid.magic.newsdaily.util.AssetsUtil.Companion.getCountries
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import java.io.IOException

class CountryPreferenceDialogFragment : PreferenceDialogFragmentCompat() {
    var country: String = ""

    companion object {
        private val TAG = CountryPreferenceDialogFragment::class.java.simpleName

        @JvmStatic
        fun newInstance(key: String): CountryPreferenceDialogFragment {
            val fragment = CountryPreferenceDialogFragment()
            fragment.arguments = bundleOf(ARG_KEY to key)
            return fragment
        }
    }

    override fun onBindDialogView(view: View) {
        super.onBindDialogView(view)
        val binding = PrefDialogCountryBinding.bind(view)
        try {
            val countries = getCountries(context?.assets)
            val countryAdapter = CountryAdapter()
            countryAdapter.submitList(countries)
            binding.recyclerView.apply {
                adapter = countryAdapter
                layoutManager = LinearLayoutManager(context)
                itemAnimator = DefaultItemAnimator()
                addItemDecoration(DividerItemDecoration(context, DividerItemDecoration.VERTICAL))
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    override fun onDialogClosed(positiveResult: Boolean) {
        if (positiveResult) {
            val preference = preference
            if (preference is CountryPreference)
                preference.country = country
        }
    }

    @Subscribe
    fun onCountryEvent(countryEvent: CountryEvent) {
        country = countryEvent.countryCode
        Log.d(TAG, "country clicked: $country")
        dialog?.dismiss()
        onDialogClosed(true)
    }

    override fun onResume() {
        super.onResume()
        EventBus.getDefault().register(this)
    }

    override fun onPause() {
        super.onPause()
        EventBus.getDefault().unregister(this)
    }
}